### PR TITLE
Quote path settings containing # and spaces

### PR DESCRIPTION
**Pull request by jayvdb**
This is limited to settings of type "path".
While it is limited to `#` and ` ` (space), other characters which break `shlex` could be added.

Fixes https://github.com/tox-dev/tox/issues/763 and https://github.com/tox-dev/tox/issues/924

## Contribution checklist:

(also see [CONTRIBUTING.rst](../tree/master/CONTRIBUTING.rst) for details)

- [x] wrote descriptive pull request text
- [x] added/updated test(s)
- [ ] updated/extended the documentation
- [x] added relevant [issue keyword](https://help.github.com/articles/closing-issues-using-keywords/)
      in message body
- [x] added news fragment in [changelog folder](../tree/master/docs/changelog)
  * fragment name: `<issue number>.<type>.rst` for example (588.bugfix.rst)
  * `<type>` is must be one of `bugfix`, `feature`, `deprecation`,`breaking`, `doc`, `misc`
  * if PR has no issue: consider creating one first or change it to the PR number after creating the PR
  * "sign" fragment with "by :user:`<your username>`"
  * please use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files - by :user:`superuser`."
  * also see [examples](../tree/master/docs/changelog)
- [x] added yourself to `CONTRIBUTORS` (preserving alphabetical order)

### CODE REVIEW

File: src/tox/config/__init__.py [hunk 1]
```diff
@@ -390,7 +391,7 @@ def get(self, name, default=None):
                 return os.environ.get(name, default)
             self._lookupstack.append(name)
             try:
-                res = self.reader._replace(val)
+                res = self.reader._replace(val, unquote_path=False)
```
Comment (jayvdb) [inline src/tox/config/__init__.py:394]: How quoted paths should appear in the setenv results is an aspect I havent explored yet, and would likely build on https://github.com/tox-dev/tox/pull/1691

Despite stripping quotes as early as possible, I expect there are quite a few corner cases where the quotes are going to causes new complications, which is why I have ensured the quotes are only used when `#` is in the path, which already didnt work, so corner cases breaking isnt a regression.  Adding ` ` to the quoted path chars is less clear, as that probably worked for many simple cases, and only failed in corner cases as seen at https://github.com/tox-dev/tox/issues/924.

File: src/tox/config/__init__.py [hunk 2]
```diff
@@ -1591,7 +1592,7 @@ def addsubstitutions(self, _posargs=None, **kw):
             self.posargs = _posargs
 
     def getpath(self, name, defaultpath, replace=True):
-        path = self.getstring(name, defaultpath, replace=replace)
+        path = self.getstring(name, defaultpath, replace=replace, unquote_path=True)
         if path is not None:
             toxinidir = self._subs["toxinidir"]
             return toxinidir.join(path, abs=True)
```
Comment (jayvdb) [inline src/tox/config/__init__.py:1598]: I was doing the unquoting here, before the join.  It worked well, but didnt feel tight enough.

There is still one aspect which may make this necessary, and at least needs a test case.
In addition to problematic chars in preset paths like `{toxinidir}`, the space or hash might appear in derived paths, such as `{toxinidir}/foo bar.txt`

Comment (jayvdb) [reply src/tox/config/__init__.py:1598]: Added tests for `'{toxinidir}'/'foo bar.txt'` , and I am pretty sure that the unquoting cant occur here.

File: src/tox/config/__init__.py [hunk 3]
```diff
@@ -1895,7 +1915,7 @@ def getargvlist(cls, reader, value, replace=True):
             current_command += line
 
             if is_section_substitution(current_command):
-                replaced = reader._replace(current_command, crossonly=True)
+                replaced = reader._replace(current_command, crossonly=True, unquote_path=False)
```
Comment (jayvdb) [inline src/tox/config/__init__.py:1918]: `unquote_path=False` is added liberally to aid reviewing, so all invocations of `_replace(..)` can be seen.
IMO they should removed before being merged.

File: src/tox/config/__init__.py [hunk 4]
```diff
@@ -1924,8 +1944,11 @@ def processcommand(cls, reader, command, replace=True):
                     continue
 
                 new_arg = ""
-                new_word = reader._replace(word)
-                new_word = reader._replace(new_word)
+                has_dual_quote = "''" in word
+                new_word = reader._replace(word, unquote_path=False)
+                new_word = reader._replace(new_word, unquote_path=False)
+                if not has_dual_quote:
+                    new_word = new_word.replace("''", "'")
```
Comment (jayvdb) [inline src/tox/config/__init__.py:1951]: this has the potential for changing the behaviour of any existing `tox.ini` which used quotes and also had quotes inserted via substitution.  The most obvious case would be an env var of `'` and commands using the envvar and quotes.  Then, does stripping dual quotes change the behaviour of those previous usages?  In theory, this removal would be simply merging quoted strings.  Need to think more about possible breakages.

Comment (jayvdb) [reply src/tox/config/__init__.py:1951]: Ah, this simple replace would break `\\''`.  Easy fixed with a `re`.

File: src/tox/config/__init__.py [hunk 5]
```diff
@@ -1864,6 +1878,12 @@ def _replace_substitution(self, match):
             val = self._substitute_from_other_section(sub_key)
         if callable(val):
             val = val()
+        if isinstance(val, PathBase):
+            val = str(val)
+            # XXX handle ' and " in paths
```
Comment (jayvdb) [inline src/tox/config/__init__.py:1883]: This has partially been done.
I was thinking of switching quoting char to avoid clashes with `'`.  The quoting could even be done with an invalid filename char (e.g. non-printables), as long as it makes it through the regex engine.

File: tests/unit/config/test_config.py [hunk 6]
```diff
@@ -1531,6 +1645,47 @@ def test_rewrite_simple_posargs(self, tmpdir, newconfig):
...
+            [testenv:py27]
+            args_are_paths = True
+            changedir = test#dir
+            commands = cmd1 {posargs:hello}
+        """
+        conf = newconfig([], inisource).envconfigs["py27"]
+        argv = conf.commands
+        assert argv[0] == ["cmd1", "hello"]
+
+        if sys.platform != "win32":
```
Comment (jayvdb) [inline tests/unit/config/test_config.py:1659]: Noting windows specific behaviour is happening here, cutting the dir at `#`.  May be a bug in test harness code, given that the next stanza is OK.

File: src/tox/config/__init__.py (hunk 4 above)
Comment (jayvdb) [reply src/tox/config/__init__.py]: `\\''` now fixed.  No test coverage yet.  Other obscure oddities may still exist.

File: src/tox/config/__init__.py (hunk 1 above)
Comment (jayvdb) [reply src/tox/config/__init__.py]: Tests for setting paths in `setenv` have been added.

Spaces in paths also added.

Merging #1691 first would allow for more exploration of possible edge cases using escaping.

File: docs/changelog/1700.feature.rst [hunk 7]
```diff
@@ -0,0 +1 @@
+Allow {/} to refer to os.sep. - by :user:`jayvdb`
```
Comment (jayvdb) [inline docs/changelog/1700.feature.rst:1]: This is https://github.com/tox-dev/tox/pull/1701

Without this, the tests need to do hacks to convert `/` to `\` on Windows.  It gets messy because the bugs related to this are about real native paths, including constructed path based on preset paths.

**Review by gaborbernat — CHANGES_REQUESTED** [review]
Can you rebase this?

**Comment by jayvdb** [comment]
Sure, but now that #1691 is merged (thx), I'd like to also explore how all of these chars appear in the commands subprocess runtime environment variables.
Also now worth verifying that `{` in paths doesnt do anything odd.

**Comment by jayvdb** [comment]
https://github.com/tox-dev/tox/pull/1713 solves many of these issues, in as much as that `type=path` args pass through without escaping except anything inside `{..}`.  There are a few tests from this PR which were omitted, related to round tripping paths through `setenv`, but it feels it can solve the underlying issues much cleaner and clearer (and thus easier to document), but will have some problems unsolved.

**Comment by gaborbernat** [comment]
@jayvdb this is merge conflicting.

**Comment by gaborbernat** [comment]
I'll close this for now until you can attend to the PR again. Once you've done we can reopen.
